Ascs decouple sessions

fredrikdanebjer opened the pull request:
I think some refactor and changes are needed on this in adjustments to https://github.com/zephyrproject-rtos/zephyr/pull/51513 when merged.

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -47,24 +48,77 @@
 struct bt_ascs {
 	struct bt_conn *conn;
-	struct bt_ascs_ase ases[ASE_COUNT];
+	struct bt_ascs_ase *ases[ASE_COUNT];
```
Did you consider to use `sys_slist` instead of fixed-size array of pointers?

Thalley replied:
Unresolving this as I have the same question :) Did you get an answer?

fredrikdanebjer replied:
Sorry, I missed this comment initially. I changed the implementation to use `sys_slist` instead

MariuszSkamra commented on subsys/bluetooth/audio/endpoint.h:
```diff
@@ -59,23 +59,10 @@ struct bt_audio_ep {
 	struct bt_codec_qos qos;
 	struct bt_codec_qos_pref qos_pref;
 	struct bt_audio_iso *iso;
-	struct bt_gatt_subscribe_params subscribe;
-	struct bt_gatt_discover_params discover;
 
 	/* FIXME: Replace with metastate */
 	bool receiver_ready;
 
-	/* TODO: Consider client/server container split */
-	union {
-		struct {
-			uint16_t handle;
-			uint16_t cp_handle;
-		} client;
-		struct {
-			const struct bt_gatt_attr *attr;
-		} server;
-	};
-
```
Nice that you moved those to role-dedicated containers!

MariuszSkamra reviewed (changes requested):
LGTM. I have few comments or rather suggestions. Also, please fix the indents that are incorrect in few places.

Thalley commented on subsys/bluetooth/audio/Kconfig.ascs:
```diff
@@ -35,4 +35,13 @@ config BT_ASCS_ASE_SNK
+
+config BT_ASCS_MAX_ACTIVE_ASES
+    int "Number of simultaneously supported ASE sessions" 
+    default BT_MAX_CONN
```
I think that depends: We can in theory have more than `BT_ISO_MAX_CHAN` ASEs, as long as they are not in the streaming state. The number of ASEs for a single connection we can support will be equal to the number of `bt_audio_iso` we support, but it is also a bit tricky as we have an ASE per connection (unlike the number of `bt_audio_iso` or `BT_ISO_MAX_CHAN` which are a total, and also includes the number of broadcast ISO streams supported). 

I don't see any reason why we would restrict the number of ASEs to any value (the number of active ASEs will depend on the use case). 

Since this will add an ASE pool, I think defaulting to 1 ASE per connection we support works OK.

fredrikdanebjer replied:
The max is clearly wrong below though, it should be
(BT_ASCS_ASE_SNK_COUNT + BT_ASCS_ASE_SRC_COUNT) * BT_MAX_CONN
Which also would be what I'd prefer to put the default value to above.
Therefor I ask you @Thalley @MariuszSkamra  what you prefer the default to be

fredrikdanebjer replied:
I'm not sure how to best solve this, since as I understand we cannot use arithmetics in kconfig:

```
config BT_ASCS_MAX_ACTIVE_ASES
	int "Number of simultaneously supported ASE sessions"
	default 2       # BT_MAX_CONN defaults to 1, so we default to two active ases per connection
	range 1 127500  # Highest range of: (BT_MAX_CONN * (BT_ASCS_ASE_SNK_COUNT + BT_ASCS_ASE_SRC_COUNT))
	help
	  The number of simultanesouly supported ASE sessions, in particular
	  meaning the number of ASE State Machines that are allowed to be in a
	  non-IDLE state at a single time.
```

Thalley replied:
We should also consider if we are overengineering this :D While we technically can have more ASEs "active" than we support ISO channels, once an ASE enters the QoS configured state it needs an audio_iso (which is limited to the number of ISO chans we support), so we could probably also just put the number of active ASEs to be the same as number of active ISO chans. 

The only use case for keeping an ASE only in the codec configured state, is to support the caching, but then we need to use the max of number of conn and bonds we support.

fredrikdanebjer replied:
Yes, this is a really good point.

Also, we don't really support caching today if I interpret it correct. Unless the peer has disconnected us, we will always disconnects and go to IDLE.

fredrikdanebjer replied:
With that in mind I set the default, and max to BT_ISO_MAX_CHAN. As long as we don't support caching that makes sense to me.

Thalley replied:
Just keep in mind that `BT_ISO_MAX_CHAN` may be larger than `BT_MAX_CONN * (BT_ASCS_ASE_SNK_COUNT + BT_ASCS_ASE_SRC_COUNT)` :D

fredrikdanebjer replied:
I added a BUILD_ASSERT for this expressions instead, and changed the range to an integer literal.

Thalley commented:
@fredrikdanebjer Do you want my review on this PR in its current state, or is it better to wait until it has been rebased on top of https://github.com/zephyrproject-rtos/zephyr/pull/51513?

fredrikdanebjer commented:
> @fredrikdanebjer Do you want my review on this PR in its current state, or is it better to wait until it has been rebased on top of #51513?

@Thalley  I think it makes more sense to wait until the rebase!

Thalley reviewed (changes requested):
@fredrikdanebjer please rebase :)

fredrikdanebjer commented:
> @fredrikdanebjer please rebase :)
@thalley
Im working on it right now, thanks for your great patience on time on this one :)

fredrikdanebjer commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -47,24 +48,77 @@
 	struct bt_audio_ep ep;
+	const struct bt_gatt_attr *attr;
 	struct k_work work;
+	atomic_t ref;
```
Good suggestion, I have done so also :)

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -814,17 +882,21 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
-			/* ase_process will handle the final state transition into idle state */
-			ase_release(ase);
-		}
+		if (session->ases[i]) {
+			struct bt_ascs_ase *ase = session->ases[i];
+				struct bt_audio_stream *stream = ase->ep.stream;
```
indent

MariuszSkamra replied:
Would look cleaner if you first declare the variables

```
struct bt_ascs_ase *ase = &session->ases[i];
struct bt_audio_stream *stream;

if (ase == NULL)  {
   continue;
}

stream = ase->ep.stream;
```

Plus, you won't need to change the code below (so that the history is unchanged)

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.ascs:
```diff
@@ -35,4 +35,13 @@ config BT_ASCS_ASE_SNK
+
+config BT_ASCS_MAX_ACTIVE_ASES
+	int "Number of simultaneously supported ASE sessions"
+	default BT_MAX_CONN
+	range 1 BT_MAX_CONN
```
Oh, ignore this one as it's already discussed here https://github.com/zephyrproject-rtos/zephyr/pull/51699#discussion_r1008355222

fredrikdanebjer committed:
Bluetooth: Audio: Endpoint: Decouple Client and Server

Refactored out the client and server members of the endpoint struct
and placed the relevant members in wrappers in ascs and
unicast_client.

Signed-off-by: Fredrik Danebjer <fredrik@danebjer.com>

fredrikdanebjer committed:
Bluetooth: Audio: ASCS: Make ASE Allocation Dynamic

Decoupled the ASEs from the ASCS Session and made them dynamically
allocated instead. A Kconfig option was added to set the maximum
number of active ASEs at a single time. The intent here is to allow
the developer greater control over memory usage; this fix addresses
one of the largest ram usages in le audio.

Signed-off-by: Fredrik Danebjer <fredrik@danebjer.com>

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -51,17 +55,57 @@ LOG_MODULE_REGISTER(bt_ascs, CONFIG_BT_ASCS_LOG_LEVEL);
+	struct bt_ascs_ase *ases[ASE_COUNT];
 };
 
+K_MEM_SLAB_DEFINE(ase_slab, sizeof(struct bt_ascs_ase),
+		  CONFIG_BT_ASCS_MAX_ACTIVE_ASES, __alignof__(struct bt_ascs_ase));
```
```suggestion
K_MEM_SLAB_DEFINE(ase_slab, sizeof(struct bt_ascs_ase),
		  CONFIG_BT_ASCS_MAX_ACTIVE_ASES, 
		  __alignof__(struct bt_ascs_ase));
```

Recommended maximum line length is 80 characters

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -51,17 +55,57 @@ LOG_MODULE_REGISTER(bt_ascs, CONFIG_BT_ASCS_LOG_LEVEL);
+{
+	__ASSERT(ase && ase->ascs, "Non-existing ASE or ASCS");
+
+	for (uint8_t i = 0; i < ASE_COUNT; i++) {
```
```suggestion
	for (int i = 0; i < ASE_COUNT; i++) {
```

`ASE_COUNT` is an int

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -51,17 +55,57 @@ LOG_MODULE_REGISTER(bt_ascs, CONFIG_BT_ASCS_LOG_LEVEL);
+
+	for (uint8_t i = 0; i < ASE_COUNT; i++) {
+		if (ase->ascs->ases[i] == ase) {
+			LOG_DBG("Deactivating ASE at index=%u", i);
```
```suggestion
			LOG_DBG("Deactivating ASE at index=%d", i);
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -51,17 +55,57 @@ LOG_MODULE_REGISTER(bt_ascs, CONFIG_BT_ASCS_LOG_LEVEL);
+		}
+	}
+
+	__ASSERT(0, "Invalid ASE access");
```
```suggestion
	__ASSERT(false, "Invalid ASE access");
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -104,6 +148,10 @@ void ascs_ep_set_state(struct bt_audio_ep *ep, uint8_t state)
 			if (ops->released != NULL) {
 				ops->released(stream);
 			}
+			struct bt_ascs_ase *ase = CONTAINER_OF(ep, struct bt_ascs_ase, ep);
```
Move variable declaration to top of scope

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -352,6 +400,26 @@ static void ascs_ep_get_status_enable(struct bt_audio_ep *ep,
+		return -EINVAL;
+	}
+
+	struct bt_ascs_ase_status *status;
```
Move declaration to top of scope

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -791,8 +859,8 @@ static void ase_disable(struct bt_ascs_ase *ase)
 
-	/* The ASE state machine goes into different states from this operation
-	 * based on whether it is a source or a sink ASE.
+	/** The ASE state machine goes into different states from this operation
+	 *  based on whether it is a source or a sink ASE.
```
```suggestion
	/* The ASE state machine goes into different states from this operation
	 * based on whether it is a source or a sink ASE.
```

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -812,17 +880,27 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
 	for (size_t i = 0; i < ARRAY_SIZE(session->ases); i++) {
-		struct bt_ascs_ase *ase = &session->ases[i];
-		struct bt_audio_stream *stream = ase->ep.stream;
+		if (session->ases[i]) {
```
Is there any case where `ases[0]` may be NULL and `ases[1]` isn't? In other words, can we break this loop when we meet a NULL `session->ases[i]`, or do we need to iterate on all of them?

fredrikdanebjer replied:
Actually we do. If at some point two ASEs are allocated for a connection, index  0 and 1, then only index 0 was deallocated through a Release operation we would still have to handle index 1.

Maybe this also speaks for a sys_slist solution instead, as was commented earlier I saw :)

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -985,61 +1063,47 @@ void ascs_ep_init(struct bt_audio_ep *ep, uint8_t id)
-		i = id;
-		ase = &ascs->ases[i - 1];
-		goto done;
+	if (!id || id > ASE_COUNT) {
+		return NULL;
```
So the change here is that `ase_new` should never be called with `id == 0`, right? Previously implementation used `id == 0` to get a new ASE, but that's no longer the case. 

If we never want to call this with `id == 0`, I suggest to change this check to an __ASSERT instead, as returning NULL would probably just make it more confusing to use.

fredrikdanebjer replied:
Actually, the ase_new with id 0 was previously never called. We discussed this on discord previously.
The client has to call the config with an id, and if its not already allocated that would then have happened in ase_get with a proper id already.

In regards to ASSERT: We would only return a NULL if we dont have enough memory, in which case we should respond with a BT_ASCS_RSP_NO_MEM. This is what happens now when ase_new returns NULL to ase_get, which forwards the NULL to ascs_config.

If the client calls with in invalid ASE, this is already handled in ascs_config, where we then respondBT_ASCS_RSP_INVALID_ASE

Thalley replied:
Right, that is actually my point. Since we do the check for `id == 0` in `ascs_config`, where it isn't allowed, we don't need it here. 

Since `ase_get` may also be called internally, outside of `ascs_config`, then we should ensure that it is never 0. We could do that with this check too, but since it would only happen in case of a programming error on our side, an ASSERT might be more useful.

fredrikdanebjer replied:
I see the redundancy now.
However, `ase_get` is (no longer at least) called anywhere else. I will remove this check completely.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1355,16 +1428,19 @@ static ssize_t ascs_config(struct bt_ascs *ascs, struct net_buf_simple *buf)
-		if (cfg->ase) {
-			ase = ase_get(ascs, cfg->ase);
+		if (!cfg->ase || cfg->ase > ASE_COUNT) {
+			LOG_WRN("Invalid ASE ID");
```
```suggestion
			LOG_WRN("Invalid ASE ID: %u", cfg->ase);
```

fredrikdanebjer replied:
Ill fix

Thalley commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -227,11 +238,14 @@ static struct bt_iso_chan_ops unicast_client_iso_ops = {
 static void unicast_client_ep_init(struct bt_audio_ep *ep, uint16_t handle,
 				   uint8_t dir)
 {
+	struct bt_unicast_client_ep *client;
```
suggest to use `client_ep` instead of `client`, as `client` would be more fitting for something broader than a single remote ASE

Thalley replied:
This comment applies multiple places

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -39,6 +39,14 @@ LOG_MODULE_REGISTER(bt_ascs, CONFIG_BT_ASCS_LOG_LEVEL);
+
+BUILD_ASSERT(CONFIG_BT_ASCS_MAX_ACTIVE_ASES <= MAX(MAX_ASES_SESSIONS),
+						   CONFIG_BT_ISO_MAX_CHAN),
+	     "Too many allocated ASEs");
```
```suggestion
	     "Active ASEs %u larger than actual number of ASEs (%u) or ISOs (%u)", CONFIG_BT_ASCS_MAX_ACTIVE_ASES, MAX_ASES_SESSIONS, CONFIG_BT_ISO_MAX_CHAN);
```

fredrikdanebjer replied:
I have made the message more informative, but we cannot use format string for BUILD_ASSERT as I understand

Thalley replied:
Ah, right, forgot about that part :D
